File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    answer = 0
    n = len(numbers)
    # (현재까지 계산 결과, numbers에서 몇번째 인덱스 까지 사용했는지)
    q = deque([(numbers[0], 0), (-numbers[0], 0)])
    while q:
        result, idx = q.popleft()
        # numbers의 모든인덱스를 다 거쳐갔다면 목표숫자와 같은지 확인
        if idx == n - 1:
            if result == target:
                answer += 1
            continue
        # 다음 인덱스의 값을 + or - 추가해주기
        q.append((result + numbers[idx + 1], idx + 1))
        q.append((result - numbers[idx + 1], idx + 1))
        
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int solution(int[] numbers, int target) {
        int answer = 0;
        
        Deque<int[]> dq = new ArrayDeque<int[]>();
        // {현재 합산 숫자, 지금까지 연산을 진행한 숫자 인덱스}
        dq.offerLast(new int[]{0, 0});
        while(!dq.isEmpty()) {
            int[] sumArray = dq.pollFirst();
            
            // 모든 numbers를 사용하여 연산이 끝나면 해당 수가 타겟 넘버인지 확인
            if(sumArray[1] >= numbers.length) {
                if(sumArray[0] == target) answer++;
                continue;
            }
            
            // 다음 연산 대상 숫자를 더하거나 빼서 큐애 삽입
            dq.offerLast(new int[]{sumArray[0] + numbers[sumArray[1]], sumArray[1] + 1});
            dq.offerLast(new int[]{sumArray[0] - numbers[sumArray[1]], sumArray[1] + 1});
        }
        
        return answer;
    }
}